gbuiltins/bls381/src/lib.rs

```diff
@@ -0,0 +1,240 @@
+    ///
+    /// Encoded:
+    ///   - `base`: `ArkScaleProjective<G2Projective>`.
+    ///   - `scalar`: `ArkScale<Vec<u64>>`.
+    #[codec(index = 5)]
+    ProjectiveMultiplicationG2 { base: Vec<u8>, scalar: Vec<u8> },
+}
+
+/// The enumeration represents possible common errors for all requests.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[codec(crate = codec)]
+pub enum Error {
+    /// Failed to scale-decode the length of a `Vec<u8>`.
+    DecodeVecLength,
+    /// Failed to scale-decode bytes.
+    DecodeVecData,
+    /// Failed to decode the length of a `Vec<G(1,2)Affine>`.
+    DecodeItemCount,
+}
+
+/// The enumeration contains result to a request.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq, derive_more::From)]
+#[codec(crate = codec)]
+pub enum Response {
+    /// Common error.
+    #[from]
+    Error(Error),
+    /// Result of the multi Miller loop [`MultiMillerLoopResult`].
+    #[from]
+    MultiMillerLoop(MultiMillerLoopResult),
+    /// Result of the final exponentiation.
+    FinalExponentiation(Result<Vec<u8>, ()>),
+    /// Result of the multi scalar multiplication [`MultiScalarMultiplicationResult`].
+    #[from]
+    MultiScalarMultiplication(MultiScalarMultiplicationResult),
+    /// Result of the projective multiplication.
+    ProjectiveMultiplication(Result<Vec<u8>, ()>),
+}
+
+/// Result of the multi Miller loop computation.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[codec(crate = codec)]
+pub enum MultiMillerLoopResult {
+    /// Encoded: `ArkScale<Bls12_381::TargetField>`.
+    Ok(Vec<u8>),
+    /// Computation error.
+    Error,
+    /// The input lists don't have the same number of items.
+    NonEqualItemCount,
+}
+
+impl From<Result<Vec<u8>, ()>> for MultiMillerLoopResult {
+    fn from(result: Result<Vec<u8>, ()>) -> Self {
+        match result {
+            Ok(v) => MultiMillerLoopResult::Ok(v),
+            Err(_) => MultiMillerLoopResult::Error,
+        }
+    }
+}
+
+/// Result of the multi scalar multiplication.
+#[derive(Encode, Decode, Clone, Debug, PartialEq, Eq)]
+#[codec(crate = codec)]
+pub enum MultiScalarMultiplicationResult {
+    // Encoded: `ArkScaleProjective<ark_bls12_381::G(1,2)Projective>`.
+    Ok(Vec<u8>),
+    /// Computation error.
+    Error,
+    /// The input lists don't have the same number of items.
+    NonEqualItemCount,
+}
+
+impl From<Result<Vec<u8>, ()>> for MultiScalarMultiplicationResult {
+    fn from(result: Result<Vec<u8>, ()>) -> Self {
+        match result {
+            Ok(v) => MultiScalarMultiplicationResult::Ok(v),
+            Err(_) => MultiScalarMultiplicationResult::Error,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate alloc;
+
+    use super::*;
+    use alloc::vec;
+
+    // The standard Decode implementation cannot be used for precise gas charging.
+    // The following test checks that scale codec indexes of variants are set correctly.
+    #[test]
+    fn codec_enum_indexes() {
```

Comment:
whoa, please, impl closure to be used)) for each case
and you need to write a test or extend this about these constants matching some specific number, as for backward compatibility and modifying by accident